File: backend/aws_lambda_iot_telemetry_handler.py
"""
AWS LAMBDA FUNCTION: IoT Telemetry to PostgreSQL RDS Ingestion Handler
Runtime: Python 3.10 / 3.11 / 3.12

AWS IoT Rule SQL:
  SELECT *, topic() as topic, timestamp() as msg_timestamp FROM 'devices/+/telemetry'
  (or SELECT * FROM 'devices/#' to capture status & telemetry)

Environment Variables in AWS Lambda:
  DB_HOST = safeshunt-db.c5oesqouwl70.ap-south-1.rds.amazonaws.com
  DB_NAME = safeshunt_db
  DB_USER = postgres
  DB_PASSWORD = pisolve123
  DB_PORT = 5432
"""

import os
import json
import time
import psycopg2
from psycopg2.extras import Json

DB_HOST = os.environ.get("DB_HOST", "safeshunt-db.c5oesqouwl70.ap-south-1.rds.amazonaws.com")
DB_NAME = os.environ.get("DB_NAME", "safeshunt_db")
DB_USER = os.environ.get("DB_USER", "postgres")
DB_PASSWORD = os.environ.get("DB_PASSWORD", "pisolve123")
DB_PORT = int(os.environ.get("DB_PORT", "5432"))

def derive_paired_device(device_id, payload):
    # Check explicit field first
    if payload.get("paired_tx_id"):
        return payload["paired_tx_id"]
    if payload.get("paired_rx_id"):
        return payload["paired_rx_id"]
    if payload.get("paired_device"):
        return payload["paired_device"]

    # Match numeric suffix (e.g. TX-03 <-> RX-03, TX-01 <-> RX-01)
    if device_id.startswith("TX-") or device_id.startswith("DE-"):
        num = device_id.split("-")[1]
        return f"RX-{num}"
    elif device_id.startswith("RX-") or device_id.startswith("LD-"):
        num = device_id.split("-")[1]
        return f"TX-{num}"
    elif device_id.startswith("TX"):
        return device_id.replace("TX", "RX")
    elif device_id.startswith("RX"):
        return device_id.replace("RX", "TX")
    return "TX-01" if device_id.startswith("RX") else "RX-01"

def lambda_handler(event, context):
    print("Incoming AWS IoT Telemetry Event:", json.dumps(event, default=str))

    # Extract device ID
    device_id = event.get("deviceId") or event.get("device_id") or event.get("deviceCode") or event.get("SerialNumber")
    topic = event.get("topic", f"devices/{device_id}/telemetry" if device_id else "devices/unknown/telemetry")

    if not device_id and "topic" in event:
        parts = event["topic"].split("/")
        if len(parts) >= 2 and parts[0] == "devices":
            device_id = parts[1]

    if not device_id:
        print("ERROR: No deviceId found in event or topic.")
        return {"statusCode": 400, "body": "Missing deviceId"}

    # Extract metrics
    diagnostics = event.get("diagnostics", {}) if isinstance(event.get("diagnostics"), dict) else {}
    readings = event.get("readings", {}) if isinstance(event.get("readings"), dict) else {}
    
    battery = diagnostics.get("battery_pct") if diagnostics.get("battery_pct") is not None else (event.get("battery_level") or event.get("battery"))
    signal = diagnostics.get("gsm_rssi") if diagnostics.get("gsm_rssi") is not None else (event.get("signal_rssi") or event.get("gsm_rssi") or event.get("signal"))
    lat = event.get("latitude") or event.get("lat")
    lng = event.get("longitude") or event.get("lng")
    
    # Distance in cm
    distance_cm = readings.get("distance_cm")
    if distance_cm is None:
        distance_cm = event.get("distance_cm")
    if distance_cm is None and event.get("distance") is not None:
        try:
            distance_cm = round(float(event["distance"]) * 100)
        except Exception:
            distance_cm = None

    speed_kmh = event.get("speed_kmh") or event.get("speed")
    product_type = event.get("productType") or event.get("product_type") or ("TRANSMITTER" if device_id.startswith("TX") else "RECEIVER")

    conn = None
    try:
        conn = psycopg2.connect(
            host=DB_HOST,
            database=DB_NAME,
            user=DB_USER,
            password=DB_PASSWORD,
            port=DB_PORT,
            sslmode="require"
        )
        cur = conn.cursor()

        # 1. Insert into device_telemetry
        insert_telemetry_sql = """
            INSERT INTO device_telemetry (
                device_id, topic, payload, battery_level, signal_rssi, latitude, longitude, distance_cm, speed_kmh, recorded_at
            ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, CURRENT_TIMESTAMP)
            RETURNING id;
        """
        cur.execute(insert_telemetry_sql, (
            device_id,
            topic,
            Json(event),
            battery,
            signal,
            lat,
            lng,
            distance_cm,
            speed_kmh
        ))
        telemetry_id = cur.fetchone()[0]
        print(f"Saved to device_telemetry with ID: {telemetry_id}")

        # 2. Upsert into device_registry
        upsert_registry_sql = """
            INSERT INTO device_registry (
                device_id, device_name, serial_number, product_type, hardware_version, firmware_version, health_status, last_reading_timestamp, updated_at
            ) VALUES (
                %s, %s, %s, %s, '1.0', '1.0.0', 'ONLINE', CURRENT_TIMESTAMP, CURRENT_TIMESTAMP
            )
            ON CONFLICT (device_id) DO UPDATE SET
                health_status = 'ONLINE',
                last_reading_timestamp = CURRENT_TIMESTAMP,
                updated_at = CURRENT_TIMESTAMP;
        """
        cur.execute(upsert_registry_sql, (device_id, device_id, device_id, product_type))

        # 3. Handle Pairing and Live Session State Machine
        event_type = event.get("event")
        status = event.get("status")
        paired_device = derive_paired_device(device_id, event)
        rx_id = device_id if (device_id.startswith("RX") or device_id.startswith("LD")) else paired_device
        tx_id = device_id if (device_id.startswith("TX") or device_id.startswith("DE")) else paired_device

        # If explicit /status topic or pair events
        if "/status" in topic or event_type in ["PAIR_START", "PAIR_END", "UNEXPECTED_DISCONNECT"]:
            if event_type == "PAIR_START" or status == "PAIRED":
                cur.execute(
                    "SELECT id FROM shunting_sessions WHERE (rx_device_id = %s OR ld_code = %s) AND (status = 'LIVE' OR session_status = 'LIVE')",
                    (rx_id, rx_id)
                )
                existing = cur.fetchall()
                if not existing:
                    session_code = f"SES-{str(int(time.time()))[-6:]}-{rx_id}"
                    cur.execute("""
                        INSERT INTO shunting_sessions (
                            session_number, session_code, ld_code, rx_device_id, de_code, tx_device_id,
                            session_start, start_time, session_status, status, distance_trajectory, created_at, updated_at
                        ) VALUES (%s, %s, %s, %s, %s, %s, NOW(), NOW(), 'LIVE', 'LIVE', '[]'::jsonb, NOW(), NOW())
                    """, (session_code, session_code, rx_id, rx_id, tx_id, tx_id))
                    print(f"Started session: {rx_id} <--> {tx_id}")
            elif event_type == "PAIR_END" or status == "IDLE":
                final_val = event.get("final_distance_cm") or distance_cm or 0
                cur.execute("""
                    UPDATE shunting_sessions
                    SET 
                        session_end = NOW(),
                        end_time = NOW(),
                        session_status = 'COMPLETED',
                        status = 'COMPLETED',
                        final_distance_cm = %s,
                        final_placement_distance = %s / 100.0,
                        updated_at = NOW()
                    WHERE (rx_device_id = %s OR ld_code = %s OR tx_device_id = %s OR de_code = %s) 
                      AND (status = 'LIVE' OR session_status = 'LIVE')
                """, (final_val, final_val, rx_id, rx_id, tx_id, tx_id))
                print(f"Ended session for: {rx_id} / {tx_id}")
            elif event_type == "UNEXPECTED_DISCONNECT" or status == "OFFLINE":
                cur.execute("""
                    UPDATE shunting_sessions
                    SET 
                        session_end = NOW(),
                        end_time = NOW(),
                        session_status = 'TIMED_OUT',
                        status = 'TIMED_OUT',
                        manual_close_reason = 'Hardware Unexpected Disconnect / LWT',
                        updated_at = NOW()
                    WHERE (rx_device_id = %s OR ld_code = %s OR tx_device_id = %s OR de_code = %s)
                      AND (status = 'LIVE' OR session_status = 'LIVE')
                """, (rx_id, rx_id, tx_id, tx_id))
                print(f"Timed out session for: {rx_id} / {tx_id}")

        # 4. If distance reading is present, update trajectory or auto-create session
        if distance_cm is not None:
            point = json.dumps({
                "t": event.get("cloud_timestamp") or int(time.time() * 1000),
                "d_cm": distance_cm,
                "speed_kmh": speed_kmh or 0.0,
                "battery": battery,
                "signal": signal
            })

            cur.execute("""
                UPDATE shunting_sessions
                SET 
                    distance_trajectory = COALESCE(distance_trajectory, '[]'::jsonb) || %s::jsonb,
                    final_distance_cm = %s,
                    final_placement_distance = %s / 100.0,
                    minimum_distance = LEAST(COALESCE(minimum_distance, %s / 100.0), %s / 100.0),
                    updated_at = NOW()
                WHERE (tx_device_id = %s OR de_code = %s OR rx_device_id = %s OR ld_code = %s)
                  AND (status = 'LIVE' OR session_status = 'LIVE')
            """, (point, distance_cm, distance_cm, distance_cm, distance_cm, tx_id, tx_id, rx_id, rx_id))

            if cur.rowcount == 0:
                # Auto create live session for this pair
                session_code = f"SES-{str(int(time.time()))[-6:]}-{rx_id}"
                cur.execute("""
                    INSERT INTO shunting_sessions (
                        session_number, session_code, ld_code, rx_device_id, de_code, tx_device_id,
                        session_start, start_time, session_status, status, final_distance_cm, final_placement_distance,
                        minimum_distance, distance_trajectory, created_at, updated_at
                    ) VALUES (%s, %s, %s, %s, %s, %s, NOW(), NOW(), 'LIVE', 'LIVE', %s, %s / 100.0, %s / 100.0, %s::jsonb, NOW(), NOW())
                """, (session_code, session_code, rx_id, rx_id, tx_id, tx_id, distance_cm, distance_cm, distance_cm, json.dumps([json.loads(point)])))
                print(f"Auto-started live shunting session: {rx_id} <--> {tx_id}")

        # 5. Insert into legacy telemetry_data for backward compatibility
        try:
            cur.execute(
                "INSERT INTO telemetry_data (device_id, payload, recorded_at) VALUES (%s, %s, CURRENT_TIMESTAMP)",
                (device_id, Json(event))
            )
        except Exception:
            pass

        conn.commit()
        cur.close()
        return {"statusCode": 200, "body": json.dumps({"success": True, "id": telemetry_id})}
    except Exception as e:
        print("Database Error in Lambda:", str(e))
        if conn:
            conn.rollback()
        return {"statusCode": 500, "body": str(e)}
    finally:
        if conn:
            conn.close()

File: backend/aws_lambda_iot_telemetry_handler.js
/**
 * AWS LAMBDA FUNCTION: IoT Telemetry to PostgreSQL RDS Ingestion Handler
 * Runtime: Node.js 18.x or 20.x
 * 
 * AWS IoT Rule SQL:
 *   SELECT *, topic() as topic, timestamp() as msg_timestamp FROM 'devices/+/telemetry'
 *   (or SELECT * FROM 'devices/#' to capture status & telemetry)
 * 
 * Environment Variables in AWS Lambda:
 *   DB_HOST = safeshunt-db.c5oesqouwl70.ap-south-1.rds.amazonaws.com
 *   DB_NAME = safeshunt_db
 *   DB_USER = postgres
 *   DB_PASSWORD = pisolve123
 *   DB_PORT = 5432
 */

const { Client } = require('pg');

function derivePairedDevice(deviceId, payload) {
  if (payload.paired_tx_id) return payload.paired_tx_id;
  if (payload.paired_rx_id) return payload.paired_rx_id;
  if (payload.paired_device) return payload.paired_device;

  if (deviceId.startsWith('TX-') || deviceId.startsWith('DE-')) {
    const num = deviceId.split('-')[1];
    return `RX-${num}`;
  } else if (deviceId.startsWith('RX-') || deviceId.startsWith('LD-')) {
    const num = deviceId.split('-')[1];
    return `TX-${num}`;
  } else if (deviceId.startsWith('TX')) {
    return deviceId.replace('TX', 'RX');
  } else if (deviceId.startsWith('RX')) {
    return deviceId.replace('RX', 'TX');
  }
  return deviceId.startsWith('RX') ? 'TX-01' : 'RX-01';
}

exports.handler = async (event, context) => {
  console.log('Incoming AWS IoT Telemetry Event:', JSON.stringify(event, null, 2));

  // Extract Device ID
  let deviceId = event.deviceId || event.device_id || event.deviceCode || event.SerialNumber;
  const topic = event.topic || (deviceId ? `devices/${deviceId}/telemetry` : 'devices/unknown/telemetry');

  if (!deviceId && event.topic) {
    const parts = event.topic.split('/');
    if (parts.length >= 2 && parts[0] === 'devices') {
      deviceId = parts[1];
    }
  }

  if (!deviceId) {
    console.error('ERROR: No deviceId found in event or topic.');
    return { statusCode: 400, body: 'Missing deviceId' };
  }

  // Extract metrics
  const diagnostics = event.diagnostics || {};
  const readings = event.readings || {};

  const battery = diagnostics.battery_pct ?? event.battery_level ?? event.battery ?? null;
  const signal = diagnostics.gsm_rssi ?? event.signal_rssi ?? event.gsm_rssi ?? event.signal ?? null;
  const lat = event.latitude ?? event.lat ?? null;
  const lng = event.longitude ?? event.lng ?? null;

  let distanceCm = readings.distance_cm ?? event.distance_cm;
  if (distanceCm == null && event.distance != null) {
    try {
      distanceCm = Math.round(Number(event.distance) * 100);
    } catch (_) {
      distanceCm = null;
    }
  }

  const speedKmh = event.speed_kmh ?? event.speed ?? null;
  const productType = event.productType || event.product_type || (deviceId.startsWith('TX') ? 'TRANSMITTER' : 'RECEIVER');

  const client = new Client({
    host: process.env.DB_HOST || 'safeshunt-db.c5oesqouwl70.ap-south-1.rds.amazonaws.com',
    user: process.env.DB_USER || 'postgres',
    password: process.env.DB_PASSWORD || 'pisolve123',
    database: process.env.DB_NAME || 'safeshunt_db',
    port: parseInt(process.env.DB_PORT || '5432', 10),
    ssl: { rejectUnauthorized: false }
  });

  try {
    await client.connect();

    // 1. Insert Polling Log into device_telemetry
    const telemetryInsertSql = `
      INSERT INTO device_telemetry (
        device_id, topic, payload, battery_level, signal_rssi, latitude, longitude, distance_cm, speed_kmh, recorded_at
      ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, CURRENT_TIMESTAMP)
      RETURNING id, device_id, recorded_at;
    `;
    const res = await client.query(telemetryInsertSql, [
      deviceId,
      topic,
      JSON.stringify(event),
      battery,
      signal,
      lat,
      lng,
      distanceCm,
      speedKmh
    ]);

    console.log('Saved to device_telemetry with ID:', res.rows[0].id);

    // 2. Ensure device exists in device_registry and update last reading timestamp
    const registryUpsertSql = `
      INSERT INTO device_registry (
        device_id, device_name, serial_number, product_type, hardware_version, firmware_version, health_status, last_reading_timestamp, updated_at
      ) VALUES (
        $1, $1, $1, $2, '1.0', '1.0.0', 'ONLINE', CURRENT_TIMESTAMP, CURRENT_TIMESTAMP
      )
      ON CONFLICT (device_id) DO UPDATE SET
        health_status = 'ONLINE',
        last_reading_timestamp = CURRENT_TIMESTAMP,
        updated_at = CURRENT_TIMESTAMP;
    `;
    await client.query(registryUpsertSql, [deviceId, productType]);

    // 3. HARDWARE SESSION & PAIRING STATE MACHINE
    const eventType = event.event;
    const status = event.status;
    const pairedDevice = derivePairedDevice(deviceId, event);
    const rxId = (deviceId.startsWith('RX') || deviceId.startsWith('LD')) ? deviceId : pairedDevice;
    const txId = (deviceId.startsWith('TX') || deviceId.startsWith('DE')) ? deviceId : pairedDevice;

    if (topic.includes('/status') || ['PAIR_START', 'PAIR_END', 'UNEXPECTED_DISCONNECT'].includes(eventType)) {
      if (eventType === 'PAIR_START' || status === 'PAIRED') {
        const sessionCode = `SES-${Date.now().toString().slice(-6)}-${rxId}`;
        const existing = await client.query(
          "SELECT id FROM shunting_sessions WHERE (rx_device_id = $1 OR ld_code = $1) AND (status = 'LIVE' OR session_status = 'LIVE')",
          [rxId]
        );

        if (existing.rows.length === 0) {
          await client.query(`
            INSERT INTO shunting_sessions (
              session_number, session_code, ld_code, rx_device_id, de_code, tx_device_id,
              session_start, start_time, session_status, status, distance_trajectory, created_at, updated_at
            ) VALUES ($1, $1, $2, $2, $3, $3, NOW(), NOW(), 'LIVE', 'LIVE', '[]'::jsonb, NOW(), NOW())
          `, [sessionCode, rxId, txId]);
          console.log(`[Lambda] Session Started: ${rxId} <--> ${txId}`);
        }
      } else if (eventType === 'PAIR_END' || status === 'IDLE') {
        const finalVal = event.final_distance_cm ?? distanceCm ?? 0;
        await client.query(`
          UPDATE shunting_sessions
          SET 
            session_end = NOW(),
            end_time = NOW(),
            session_status = 'COMPLETED',
            status = 'COMPLETED',
            final_distance_cm = $1,
            final_placement_distance = $1 / 100.0,
            updated_at = NOW()
          WHERE (rx_device_id = $2 OR ld_code = $2 OR tx_device_id = $3 OR de_code = $3) AND (status = 'LIVE' OR session_status = 'LIVE')
        `, [finalVal, rxId, txId]);
        console.log(`[Lambda] Session Ended: ${rxId} / ${txId}`);
      } else if (eventType === 'UNEXPECTED_DISCONNECT' || status === 'OFFLINE') {
        await client.query(`
          UPDATE shunting_sessions
          SET 
            session_end = NOW(),
            end_time = NOW(),
            session_status = 'TIMED_OUT',
            status = 'TIMED_OUT',
            manual_close_reason = 'Hardware Unexpected Disconnect / LWT',
            updated_at = NOW()
          WHERE (rx_device_id = $1 OR ld_code = $1 OR tx_device_id = $2 OR de_code = $2) AND (status = 'LIVE' OR session_status = 'LIVE')
        `, [rxId, txId]);
        console.log(`[Lambda] Session Timed Out: ${rxId} / ${txId}`);
      }
    }

    // 4. Distance Telemetry Point from TX or RX
    if (distanceCm !== null && distanceCm !== undefined) {
      const point = JSON.stringify({
        t: event.cloud_timestamp || Date.now(),
        d_cm: distanceCm,
        speed_kmh: speedKmh || 0.0,
        battery: battery,
        signal: signal
      });

      const updateRes = await client.query(`
        UPDATE shunting_sessions
        SET 
          distance_trajectory = COALESCE(distance_trajectory, '[]'::jsonb) || $1::jsonb,
          final_distance_cm = $2,
          final_placement_distance = $2 / 100.0,
          minimum_distance = LEAST(COALESCE(minimum_distance, $2 / 100.0), $2 / 100.0),
          updated_at = NOW()
        WHERE (tx_device_id = $3 OR de_code = $3 OR rx_device_id = $4 OR ld_code = $4)
          AND (status = 'LIVE' OR session_status = 'LIVE')
      `, [point, distanceCm, txId, rxId]);

      if (updateRes.rowCount === 0) {
        const sessionCode = `SES-${Date.now().toString().slice(-6)}-${rxId}`;
        await client.query(`
          INSERT INTO shunting_sessions (
            session_number, session_code, ld_code, rx_device_id, de_code, tx_device_id,
            session_start, start_time, session_status, status, final_distance_cm, final_placement_distance,
            minimum_distance, distance_trajectory, created_at, updated_at
          ) VALUES ($1, $1, $2, $2, $3, $3, NOW(), NOW(), 'LIVE', 'LIVE', $4, $4 / 100.0, $4 / 100.0, $5::jsonb, NOW(), NOW())
        `, [sessionCode, rxId, txId, distanceCm, JSON.stringify([JSON.parse(point)])]);
        console.log(`[Lambda] Auto-created live session: ${rxId} <--> ${txId}`);
      }
    }

    // 5. Insert into legacy telemetry_data for compatibility
    try {
      await client.query(
        'INSERT INTO telemetry_data (device_id, payload, recorded_at) VALUES ($1, $2, CURRENT_TIMESTAMP)',
        [deviceId, JSON.stringify(event)]
      );
    } catch (_) {}

    await client.end();
    return { statusCode: 200, body: JSON.stringify({ success: true, id: res.rows[0].id }) };
  } catch (err) {
    console.error('Database Error in Lambda:', err);
    try { await client.end(); } catch (_) {}
    return { statusCode: 500, body: err.message };
  }
};
